### ISSUE TITLE
McElieceKobaraImaiCipher - When encrypting less than 165 bytes, the decrypted byte array will always contain 165 bytes 

### ISSUE THREAD

**Issue by mcarleio** [bcgit/bc-java#943]
First of, I am using JDK 11 and bcprov-ext-jdk15on in version 1.68.

When you execute below code, you will see, that the `decrypted` byte array is not equals the original `dataToEncrypt`. This is, because on decrypt, the resulting byte array always has a minimum size of 16, filled with zeros after the correctly decrypted first 100 bytes.

    import org.bouncycastle.crypto.AsymmetricCipherKeyPair;
    import org.bouncycastle.crypto.params.AsymmetricKeyParameter;
    import org.bouncycastle.pqc.crypto.mceliece.*;

    import java.security.SecureRandom;
    import java.util.Arrays;

    public class Encrypter {

        public static final SecureRandom SECURE_RANDOM;

        static {
            try {
                SECURE_RANDOM = SecureRandom.getInstance("SHA1PRNG");
            } catch (Exception nsae) {
                throw new RuntimeException(nsae);
            }
        }

        public static void main(String[] args) throws Exception {
            // Generate random data
            byte[] dataToEncrypt = new byte[100];
            SECURE_RANDOM.nextBytes(dataToEncrypt);

            // Generate key pair
            AsymmetricCipherKeyPair keyPair = generateMcElieceKeyPair(new McElieceCCA2Parameters());

            byte[] encrypted = encryptMcEliece(dataToEncrypt, keyPair.getPublic());
            byte[] decrypted = decryptMcEliece(encrypted, keyPair.getPrivate());

            System.out.println("Result: " + Arrays.equals(dataToEncrypt, decrypted));
            assert Arrays.equals(dataToEncrypt, decrypted);
        }

        public static AsymmetricCipherKeyPair generateMcElieceKeyPair(McElieceCCA2Parameters parameters) {
            McElieceCCA2KeyPairGenerator mcElieceKeyPairGenerator = new McElieceCCA2KeyPairGenerator();
            mcElieceKeyPairGenerator.init(new McElieceCCA2KeyGenerationParameters(SECURE_RANDOM, parameters));
            return mcElieceKeyPairGenerator.generateKeyPair();
        }

        private static byte[] encryptMcEliece(
                final byte[] bytesToEncrypt,
                final AsymmetricKeyParameter publicKey
        ) {
            McElieceKobaraImaiCipher cipher = new McElieceKobaraImaiCipher();
            cipher.init(true, publicKey);
            return cipher.messageEncrypt(bytesToEncrypt);
        }

        private static byte[] decryptMcEliece(
                final byte[] bytesToDecrypt,
                final AsymmetricKeyParameter privateKey
        ) throws Exception {
            McElieceKobaraImaiCipher cipher = new McElieceKobaraImaiCipher();
            cipher.init(false, privateKey);
            return cipher.messageDecrypt(bytesToDecrypt);
        }
    }

**Comment by tonywasher** [comment]
A quick view of the JCA implementation of the  McElieceKobaraImaiCipher at [github](https://github.com/bcgit/bc-java/blob/master/prov/src/main/java/org/bouncycastle/pqc/jcajce/provider/mceliece/McElieceKobaraImaiCipherSpi.java) shows that the cipher needs padding to be applied. For this particular cipher the padding is a trailing 0x1 byte.

So your encrypt/decrypt methods should look something like this `

    private static byte[] encryptMcEliece(
            final byte[] bytesToEncrypt,
            final AsymmetricKeyParameter publicKey
    ) {
        McElieceKobaraImaiCipher cipher = new McElieceKobaraImaiCipher();
        cipher.init(true, publicKey);
        byte[] paddedData = Arrays.copyOf(bytesToEncrypt, bytesToEncrypt.length + 1);
        paddedData[bytesToEncrypt.length] = 0x1;
        return cipher.messageEncrypt(paddedData);
    }

    private static byte[] decryptMcEliece(
            final byte[] bytesToDecrypt,
            final AsymmetricKeyParameter privateKey
    ) throws Exception {
        McElieceKobaraImaiCipher cipher = new McElieceKobaraImaiCipher();
        cipher.init(false, privateKey);
        byte[] myPadded = cipher.messageDecrypt(bytesToDecrypt);
        int index = myPadded.length - 1;
        while (index >= 0 && myPadded[index] == 0) index--;

        // check if padding byte is valid
        if (myPadded[index] != 0x01)
        {
            throw new BadPaddingException("invalid ciphertext");
        }
        byte[] mBytes = new byte[index];
        System.arraycopy(myPadded, 0, mBytes, 0, index);
        return mBytes;
    }`

**Comment by mcarleio** [comment]
Well, thank you, this will work. But why is it necessary for me to add it and not part of the cipher implementation itself? As far as I can see, there already seems to be a [padding check](https://github.com/bcgit/bc-java/blob/master/prov/src/main/java/org/bouncycastle/pqc/jcajce/provider/mceliece/McElieceKobaraImaiCipherSpi.java#L205) which is kind of broken: #942

**Comment by tonywasher** [comment]
Generally speaking padding is separate from the cipher operation. The confusion here is that the error refers to padding whereas the actual error relates to the fact the the integrity value is not matched - nothing to do with padding. I will comment on #942 separately

**Comment by mcarleio** [comment]
Your fix for #942 (#944) will not remove the need to add the padding bit manually, right?
Is there any reason why I have to add it here manually, but when using other ciphers like McEliecePointchevalCipher and McElieceFujisakiCipher I do not?
Maybe I am wrong, but in my understanding, a cipher should return only the decrypted data and not implementation details like padding. Otherwise I may not simply replace one cipher with another.

**Comment by tonywasher** [comment]
Correct, you will need the padding in order to ensure that the output is the correct length

Some cipher algorithms (e.g. Pointcheval/Fujisaki) will automatically recover the length of the input data. Others (eg. KobaraImai/ RSA) will not and require padding to achieve this. Its all down to how they are specified.

If you are looking for a cipher to handle implementation details like padding etc. then you should be using the JCA interface rather than the lightweight API.

**Comment by dghgit** [comment]
I think Tony has covered this. One addition - the Kobara Imai construction requires a minimum message length, if you go over that there's no need to pad, so whether you want to pad or not is an open question.

There's more details here: https://www.degruyter.com/document/doi/10.1515/JMC.2007.009/html (unfortunately the title was not quite correct in the JavaDoc, I've corrected this). Marking this as closed. We've started work on the NIST candidate so hopefully we'll have something both stable and standard for people to use soon.
